services/src/main/scala/cromwell/pipeline/service/ControllerClient.scala

```diff
@@ -0,0 +1,84 @@
+import java.nio.charset.Charset
+
+import akka.actor.ActorSystem
+import akka.http.scaladsl.model.headers.RawHeader
+import akka.http.scaladsl.model.{ HttpMethods, HttpRequest, HttpResponse }
+import akka.http.scaladsl.{ Http, HttpExt }
+import akka.stream.ActorMaterializer
+import akka.util.ByteString
+
+import scala.concurrent.{ ExecutionContext, Future }
+
+class ControllerClient extends HttpClient {
+  private implicit val actorSystem: ActorSystem = ActorSystem()
+  private implicit val materializer: ActorMaterializer = ActorMaterializer()
+
+  def http: HttpExt = Http()
+
+  override def get(url: String, params: Map[String, String] = Map(), headers: Map[String, String] = Map())(
+    implicit ec: ExecutionContext
+  ): Future[Response] = {
+    val request =
+      HttpRequest(method = HttpMethods.GET, uri = queryBuilder(url, params)).withHeaders(parseHeaders(headers))
+    val futureResponse: Future[HttpResponse] = http.singleRequest(request)
+
+    responsify(futureResponse)
+  }
+
+  override def post(
+    url: String,
+    params: Map[String, String] = Map(),
+    headers: Map[String, String] = Map(),
+    payload: String = ""
+  )(implicit ec: ExecutionContext): Future[Response] = {
+    val request = HttpRequest(method = HttpMethods.POST, uri = queryBuilder(url, params))
+      .withEntity(payload)
+      .withHeaders(parseHeaders(headers))
+    val futureResponse: Future[HttpResponse] = http.singleRequest(request)
+
+    responsify(futureResponse)
+  }
+
+  override def put(
+    url: String,
+    params: Map[String, String] = Map(),
+    headers: Map[String, String] = Map(),
+    payload: String = ""
+  )(implicit ec: ExecutionContext): Future[Response] = {
+    val request = HttpRequest(method = HttpMethods.PUT, uri = queryBuilder(url, params))
+      .withEntity(payload)
+      .withHeaders(parseHeaders(headers))
+    val futureResponse: Future[HttpResponse] = http.singleRequest(request)
+
+    responsify(futureResponse)
+  }
+
+  private def responsify(futureResponse: Future[HttpResponse])(implicit ec: ExecutionContext): Future[Response] =
+    futureResponse.flatMap { response: HttpResponse =>
+      response.entity.dataBytes.runFold(ByteString(""))(_ ++ _).map { body =>
+        val responseBody = body.decodeString(Charset.defaultCharset())
+        Response(
+          response.status.intValue(),
+          responseBody,
+          response.headers.foldLeft(Map[String, List[String]]()) {
+            case (acc, header) =>
+              acc + (header.name() -> (header.value() :: acc.getOrElse(header.name(), Nil)))
+          }
+        )
+      }
+    }
+
+  private def parseHeaders(headers: Map[String, String]): List[RawHeader] =
+    headers.foldLeft(List.empty[RawHeader]) {
+      case (acc, (param, value)) => acc ++ List(RawHeader(param, value))
+    }
+
+  private def queryBuilder(url: String, params: Map[String, String]): String = {
+    val confirmParams = params.map { case (param, value) => s"$param=$value" }
```

Comment:
use "collect" instead of "map", to avoid matcherror, when working with partial functions ( @myazinn @ilyaonishenko, please confirm or reject this, I need your advice here )

Reply:
This is a good advice in general.
However, in that case it may only confuse others. Can you give an example of the `params` ​​values when this function will throw MatchError?

Reply:
may be, this PartialFunction could be more complex, for example, it would not be defined at string.length >= 255, or with Strings which contains special characters, delimiters etc with some regex. We can intercept this mismatch here allready

Reply:
Params are Map of Strings, so no matcherror is likely, but thank you for advice, I will take it.